### ISSUE TITLE
Support for Final-as-default, like Rust

### ISSUE THREAD

**Issue by willhansen** [python/mypy#16313]
**Feature**

If I want to follow "best practices", I've got to put a `Final[  ]` wrapper around most type hints I use.  Could there be a mypy flag to assume every type hint is final unless it's got a `Mut[  ]` wrapper?

Would probably require a search-replace script to convert a project between the modes, but that doesn't sound too bad.

**Pitch**

Variables being constant when you won't need to change them is generally regarded as a good idea, right?  This would make it easier, like Rust does.

I don't know about any of the mypy internals, but this doesn't strike me as too big a change?

Interface change would be a new `--final-as-default` flag.

**Comment by sobolevn** [comment]
I love this idea :)

I think that it needs to be clarified, that the flag should be something like `--implicit-final-vars`, because `final` might also mean classes, methods, etc.

**Comment by takeda** [comment]
Yes! I'm making myself to use `Final` in my code frequently and most of the time the variables aren't really mutated, so having a mode where `Final` would be implicit would be really helpful.

**Comment by JPHutchins** [comment]
What is the scope of work for implementing this as a plugin so that it can be tested out?

**Comment by sobolevn** [comment]
plugin would be amazing to try!

**Comment by JPHutchins** [comment]
I wonder if this would get us `Final` for destructuring - or is there already a way?

```python
done, pending = await asyncio.wait(...)
```
